File: Java/Best_Time_to_Buy_and_Sell_Stock.java
/**
 * Problem Link: https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
 * 
 */
public class Solution {
    public int maxProfit(int[] prices) {
        if(prices == null || prices.length < 2){
            return 0;
        }
        int n = prices.length;
        int profit = 0;
        int[] buy = new int[n];
        for(int i=0; i<n; i++) {
            if(i==0) buy[i] = prices[i];
            else buy[i] = Math.min(prices[i], buy[i-1]);
        }
        
        for(int i=0; i<n; i++) {
            profit = Math.max(profit, prices[i] - buy[i]);
        }
        
        return profit;
    }
}

File: Swift/best-time-to-buy-and-sell-stock.swift
/**
 * https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
 * 
 * 
 */ 
class Solution {
    func maxProfit(_ prices: [Int]) -> Int {
        guard prices.count > 0 else {return 0}
        let n = prices.count
        var maxp = 0
        var profit = 0
        for i in 0..<n {
            maxp = max(maxp, prices[n-i-1])
            profit = max(profit, maxp - prices[n-i-1])
        }
        return profit
    }
}
